<launch>
<arg name="use_encoder" value="true" />
<arg name="leg_detection" value="true" />
<arg name="torso_detection" value="true" />
<arg name="sound_detection" value="true" />
<arg name="gesture_detection" value="false" />
<arg name="bayesian" value="false"/>
<arg name="leg_node" value="true"/>
<arg name="sound_node" value="true"/>
<arg name="vision_node" value="true"/>
<arg name="human_node" value="true"/>
<arg name="test" value="false"/>
<arg name="angle_step" value="1"/>
<arg name="power_threshold" value="30.0"/>

<group if="$(arg bayesian)">
    <node pkg="likelihood_grid" type="likelihood_grid_node" name="likelihood_grid_node" output="screen" >

            <param name="use_sim_time" value="true" />
            <param name="LikelihoodGrid/resolution" value="0.25"/>
            <param name="LikelihoodGrid/grid_angle_min" value="-3.14"/>
            <param name="LikelihoodGrid/grid_angle_max" value="3.14"/>
            <param name="LikelihoodGrid/grid_range_min" value="0.0"/>
            <param name="LikelihoodGrid/grid_range_max" value="10.0"/>
            <param name="LikelihoodGrid/angle_step" value="$(arg angle_step)"/>

            <param name="LikelihoodGrid/update_rate" value="0.25"/>

            <param name="LikelihoodGrid/human_cell_probability" value="0.95"/>
            <param name="LikelihoodGrid/free_cell_probability" value="0.05"/>
            <param name="LikelihoodGrid/unknown_cell_probability" value="0.25"/>

            <!--leg-->
            <param name="LikelihoodGrid/leg_human_cell_probability" value="0.95"/>
            <param name="LikelihoodGrid/leg_free_cell_probability" value="0.05"/>
            <param name="LikelihoodGrid/leg_unknown_cell_probability" value="0.25"/>

            <!--torso-->

            <param name="LikelihoodGrid/torso_human_cell_probability" value="0.95"/>
            <param name="LikelihoodGrid/torso_free_cell_probability" value="0.05"/>
            <param name="LikelihoodGrid/torso_unknown_cell_probability" value="0.25"/>

            <!--sound-->

            <param name="LikelihoodGrid/sound_human_cell_probability" value="0.95"/>
            <param name="LikelihoodGrid/sound_free_cell_probability" value="0.05"/>
            <param name="LikelihoodGrid/sound_unknown_cell_probability" value="0.25"/>

            <param name="LikelihoodGrid/target_detection_probability" value="0.9"/>
            <param name="LikelihoodGrid/false_positive_probability" value="0.05"/>
            <param name="LikelihoodGrid/sensitivity" value="1"/>

            <param name="LikelihoodGrid/leg_range_stdev" value="0.1"/>
            <param name="LikelihoodGrid/leg_range_min" value="0.5"/>
            <param name="LikelihoodGrid/leg_range_max" value="10.0"/>

            <param name="LikelihoodGrid/leg_angle_stdev" value="0.1"/>
            <param name="LikelihoodGrid/leg_angle_min" value="-2.0943951"/>
            <param name="LikelihoodGrid/leg_angle_max" value="2.0943951"/>

            <param name="LikelihoodGrid/torso_range_stded" value="2.0"/>
            <param name="LikelihoodGrid/torso_range_min" value="2.0"/>
            <param name="LikelihoodGrid/torso_range_max" value="8.0"/>

            <param name="LikelihoodGrid/torso_angle_stdev" value="5.0"/>   <!-- in degree-->
            <param name="LikelihoodGrid/torso_angle_min" value="-0.610865238"/>
            <param name="LikelihoodGrid/torso_angle_max" value="0.610865238"/>


            <param name="LikelihoodGrid/sound_range_stdev" value="3.0"/>
            <param name="LikelihoodGrid/sound_range_min" value="1.0"/>
            <param name="LikelihoodGrid/sound_range_max" value="10.0"/>

            <param name="LikelihoodGrid/sound_angle_stdev" value="5.0"/>   <!-- in degree-->
            <param name="LikelihoodGrid/sound_angle_min" value="-3.14"/>
            <param name="LikelihoodGrid/sound_angle_max" value="3.14"/>

            <param name="LikelihoodGrid/periodic_range_stdev" value="1.0"/>
            <param name="LikelihoodGrid/periodic_angle_stdev" value="20.0"/>
            <param name="loop_rate" value="10"/>
            <param name="motion_model_enable" value="$(arg use_encoder)"/>
            <param name="leg_detection_enable" value="$(arg leg_detection)"/>
            <param name="torso_detection_enable" value="$(arg torso_detection)"/>
            <param name="sound_detection_enable" value="$(arg sound_detection)"/>
            <param name="periodic_gesture_detection_enable" value="$(arg gesture_detection)"/>

            <param name="fuse_multiply" value="false"/>
    	    <!-- <remap from="/legs" to="" /> -->
            <remap from="/torso" to="person_detection/people" />
            <remap from="/sound" to="HarkSource" />
            <remap from="/gesture" to="periodic_gestures/raw_detections" />
            <remap from="/encoder" to="/husky/odom"/>
        </node>
</group>

<group if="$(arg leg_node)">
        <node pkg="likelihood_grid" type="leg_grid_node" name="leg_grid_node" output="screen">
            <param name="leg/projection_angle_step" value="$(arg angle_step)"/>
            </node>
</group>

<group if="$(arg sound_node)">
        <node pkg="likelihood_grid" type="sound_grid_node" name="sound_grid_node" output="screen">
            <remap from="sound" to="HarkSource" />
            <param name="sound/projection_angle_step" value="$(arg angle_step)"/>
            <param name="sound/power_threshold" value="$(arg power_threshold)"/>
        </node>
</group>

<group if="$(arg vision_node)">
        <node pkg="likelihood_grid" type="vision_grid_node" name="vision_grid_node" output="screen" >
            <remap from="torso" to="person_detection/people" />
            <param name="vision/projection_angle_step" value="$(arg angle_step)"/>
        </node>
</group>

<group if="$(arg human_node)">
        <node pkg="likelihood_grid" type="human_grid_node" name="human_grid_node" output="screen">
            <param name="leg_enable" value="$(arg leg_node)"/>
            <param name="torso_enable" value="$(arg vision_node)"/>
            <param name="sound_enable" value="$(arg sound_node)"/>
            <param name="human/projection_angle_step" value="$(arg angle_step)"/>
            <rosparam>
                leg_weight: 2
                sound_weight: 4
                torso_weight: 3
            </rosparam>
        </node>
</group>

<group if="$(arg leg_detection)">
    <include file="$(find autonomy_leg_detection)/launch/leg_detection.launch"/>
</group>

<group if="$(arg torso_detection)">
    <include file="$(find person_detection)/launch/person_detection.launch"/>
    <remap from="axis/image_raw/decompressed" to="/camera/front/rgb/image_color/compressed" />
</group>


<group if="$(arg test)">
    <node pkg="likelihood_grid" type="test_node" name="test_node"/>
</group>
</launch>
